util: Get rid of RecursiveMutex in Get{Blocks,Data}Dir

hebasto opened the pull request:
Another step on the [way](https://github.com/bitcoin/bitcoin/pull/19180) to replacing all of the `RecursiveMutex` instances with the `Mutex` ones.

This PR removes the `RecursiveMutex` object by splitting it into two `Mutex` objects, and ensuring they are always acquired in the same order.

Related to #19303.

vasild commented on src/util/system.cpp:
```diff
@@ -578,75 +577,82 @@ fs::path GetDefaultDataDir()
+static fs::path pathCachedNetSpecific GUARDED_BY(g_path_cache_mutex);
 
-const fs::path &GetBlocksDir()
+namespace {
```
Why is this `namespace` required?

hebasto replied:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Rs-unnamed2

vasild reviewed (approved):
ACK f7f41a08

hebasto committed:
util: Get rid of RecursiveMutex in Get{Blocks,Data}Dir

This change removes the RecursiveMutex object by splitting it into two
Mutex objects, and ensuring they are always acquired in the same order.

hebasto committed:
scripted-diff: Rename datadir path variables in util/system.cpp

-BEGIN VERIFY SCRIPT-
sed -i 's/pathCachedNetSpecific/g_datadir_path_cached_net_specific/' src/util/system.cpp
sed -i 's/pathCached/g_datadir_path_cached/' src/util/system.cpp
-END VERIFY SCRIPT-

hebasto commented:
Reworked without introducing a new helper function.

ping @vasild @promag @MarcoFalke

vasild commented on src/util/system.cpp:
```diff
@@ -686,14 +686,12 @@ fs::path StripRedundantLastElementsOfPath(const fs::path& path)
 const fs::path &GetBlocksDir()
 {
-    LOCK(csPathCached);
+    LOCK(g_blocksdir_path_mutex);
     fs::path &path = g_blocks_path_cache_net_specific;
```
Assuming `g_blocksdir_path_mutex` protects just the variable `g_blocks_path_cache_net_specific`, would it be possible to release the mutex after this assignment?

If "yes", then I would suggest to do it as a good practice of minimizing the time a mutex is held and also to avoid holding it together with the other newly introduced `g_datadir_path_mutex`.

If "no" then I think it is warranted to explain in a comment next to `g_blocksdir_path_mutex` what exactly is protected by it.

hebasto replied:
> Assuming `g_blocksdir_path_mutex` protects just the variable `g_blocks_path_cache_net_specific`, would it be possible to release the mutex after this assignment?

I guess not, as the `path` references to the same memory location as the `g_blocks_path_cache_net_specific` variable, that must be protected by the `g_blocksdir_path_mutex`.

hebasto replied:
> If "no" then I think it is warranted to explain in a comment next to `g_blocksdir_path_mutex` what exactly is protected by it.

I think `GUARDED_BY` is pretty self-documented annotation, no?

vasild replied:
oh well, I overlooked the fact that it is being assigned to a reference

vasild reviewed (commented):
Looks good.

The commit message reads just

> util: Get rid of RecursiveMutex in Get{Blocks,Data}Dir

maybe it would be good to explain that we remove the recursive mutex by splitting it to two mutexes and ensuring they are always acquired in the same order.

hebasto commented:
Updated 0173e6884568bf9f03d08b80ffc0b80729378e44 -> 7b90f77146d36ef21760f2ce3704e2f2f463b3a2 ([pr19213.02](https://github.com/hebasto/bitcoin/commits/pr19213.02) -> [pr19213.03](https://github.com/hebasto/bitcoin/commits/pr19213.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19213.02..pr19213.03)):

- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19213#pullrequestreview-526095156):
> Looks good.
> 
> The commit message reads just
> 
> > util: Get rid of RecursiveMutex in Get{Blocks,Data}Dir
> 
> maybe it would be good to explain that we remove the recursive mutex by splitting it to two mutexes and ensuring they are always acquired in the same order.

vasild reviewed (approved):
ACK 7b90f7714

vasild commented:
Why close?

hebasto commented:
> Why close?

It is incompatible with #21244.
